sstables/sstable_set.cc

```diff
@@ -464,6 +470,105 @@ class time_series_sstable_set : public sstable_set_impl {
     }
 };
 
+// Reader of a sstable together with its `max_position` (see `sstable::max_position`).
+struct reader_and_positions {
+    flat_mutation_reader reader;
+    position_in_partition max_position;
+
+    reader_and_positions(flat_mutation_reader r, position_in_partition max_pos)
+        : reader(std::move(r)), max_position(std::move(max_pos)) {}
+};
+
+// Queue of readers of sstables in a time_series_sstable_set,
+// returning readers in order of the sstables' min_position()s.
+//
+// Skips sstables that don't pass the supplied filter.
+//
+// The readers are created lazily on-demand using the supplied factory function.
+class min_position_reader_queue {
+    using container_t = time_series_sstable_set::container_t;
+    using value_t = container_t::value_type;
+
+    schema_ptr _schema;
+    lw_shared_ptr<const container_t> _sstables;
+
+    // Iterates over sstables in order of min_position().
+    container_t::const_iterator _it;
+    const container_t::const_iterator _end;
+
+    position_in_partition::tri_compare _cmp;
+
+    std::function<flat_mutation_reader(sstable&)> _create_reader;
+    std::function<bool(const sstable&)> _filter;
+
+    flat_mutation_reader create_reader(sstable& sst) {
+        return _create_reader(sst);
+    }
+
+    bool filter(const sstable& sst) const {
+        return _filter(sst);
+    }
+
+public:
+    min_position_reader_queue(schema_ptr schema,
+            lw_shared_ptr<const time_series_sstable_set::container_t> sstables,
+            std::function<flat_mutation_reader(sstable&)> create_reader,
+            std::function<bool(const sstable&)> filter)
+        : _schema(std::move(schema))
+        , _sstables(std::move(sstables))
+        , _it(_sstables->begin())
+        , _end(_sstables->end())
```

Comment:
Can these iterators be invalidated if the set is mutated while the read is going on?

Reply:
Yes, but AFAIU sstable_sets are not mutated after they start being used; the `insert` and `erase` functions are only for creating the sstable sets, and if one wants a new set with a new sstable inserted, for example, one must do `clone` + `insert` on the new set.

I think `partitioned_sstable_set::incremental_selector` uses the same assumption (it stores a `map_iterator` to `_leveled_sstables` which is a reference to `partitioned_sstable_set::_leveled_sstables`).

Correct me if I'm wrong.

Reply:
Right, I think we are OK here.